### ISSUE TITLE
How to get expectedSequenceToken from PutLogEvents response in CloudWatchLogsClient ?

### ISSUE THREAD

**Issue by researchdeveloping** [aws/aws-sdk-cpp#1371]
Confirm by changing [ ] to [x] below:
- [x] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/welcome.html) and [API reference](http://sdk.amazonaws.com/cpp/api/LATEST/index.html)
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-cpp/issues) and didn't find any solution

**Platform/OS/Hardware/Device**
Windows 10

**Describe the question**
How to get expectedSequenceToken from PutLogEvents response in CloudWatchLogsClient ? The only code documentation says 

> "You can also get the sequence token in the <code>expectedSequenceToken</code> field from<code>InvalidSequenceTokenException</code>"

. I see this token in AWSErrorMarshall trace logs but how can I access it from code ?

**Logs/output**
If applicable, add logs or error output.

`[TRACE] 2020-04-29 13:49:00.586 AWSErrorMarshaller [13316] Error response is {
	"__type":	"InvalidSequenceTokenException",
	"expectedSequenceToken":	"49606112957032105680939514520911727570644721356410751139",
	"message":	"The given sequenceToken is invalid. The next expected sequenceToken is: 49606112957032105680939514520911727570644721356410751139"
}`

**Comment by KaibaLopez** [comment]
Hi @researchdeveloping,
I think you're looking for this:
```
PutLogEventsRequest putRequest;
putRequest.WithLogGroupName("testLogGroup")
                   .WithLogStreamName("testLogStream");
//Input your logs//
auto putOutcome = m_client->PutLogEvents(putRequest);
if(putOutcome.IsSuccess())
    auto nextSeqToken = putOutcome.GetResult().GetNextSequenceToken();
```
That should return the token, but let me know if that's not what you meant.

**Comment by researchdeveloping** [comment]
Hi @KaibaLopez,

unfortunatelly, that is not what I meant (this will work in the case of successfull put new log event). But it not work in my scenario.

What I need is the "expectedSequenceToken" in the case when request fail (putOutcome.IsSuccess() will return fail) with error: "Exception name: InvalidSequenceTokenException
Error message: The given sequenceToken is invalid. The next expected sequenceToken is: XXX") and the only way to put next log is to pass this returned expectedSequenceToken for a new request. For now I know how to get if manually from log, but I need to get it from code (of course I can parse this error message to get it but this is not a good way). This scenario you will get when you will put your event next time without providing sequenceToken from previous response.

I hope this will help you to help me :)

**Comment by KaibaLopez** [comment]
oh, my bad there but yea this kind of complicates things....

So there's good and bad news on this, we don't really expose those values at the moment, BUT we are releasing version 1.8 soon, where we'd have modeled exceptions for this kind of thing, and when that happens you could do something like : 

`Aws::String token = outcome.GetError<InvalidSequenceTokenException>().GetExpectedSequenceToken();`

While that happens though and I wouldn't really recommend this but, in case this is a blocking thing for you, you can always get the error message as a string and then you can parse out the token kind of like:

```
std::cout<< putOutcome.GetError().GetMessage()<< std::endl;
auto SeqTok_loc = putOutcome.GetError().GetMessage().find("is:");
auto SeqTok_loc_end = putOutcome.GetError().GetMessage().find(" with address");
std::cout<< "token: " << putOutcome.GetError().GetMessage().substr(SeqTok_loc+4,SeqTok_loc_end-(SeqTok_loc+4) )<< std::endl;   
```   
In any case I'll comment here once the new version releases since that'd be a much better solution.
Hope this helps and let me know if you have other questions.

**Comment by researchdeveloping** [comment]
Thank you @KaibaLopez , so I will wait for update SDK. As I see there are more error classes to encapsulate and I hope you will add tchem all :) At the moment as you write I parse error message, in my case by regex expression. Your one may have better performance, but to update it I will wait for a new version of SDK to do it in correct way!

**Comment by singku-china** [comment]
@researchdeveloping We have published [SDK 1.8 preview build](https://github.com/aws/aws-sdk-cpp/tree/version1.8) which include Modeled exception feature that allows you access detailed exception fields. Check out the [blogpost](https://aws.amazon.com/blogs/developer/aws-sdk-for-c-version-1-8-developer-preview/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+AwsDeveloperBlog+%28AWS+Developer+Blog%29) and search for `modeled exceptions`, or take a look at our [Github wiki page](https://github.com/aws/aws-sdk-cpp/wiki/What%E2%80%99s-New-in-AWS-SDK-for-CPP-Version-1.8).
